source_integration_test.go

```diff
@@ -0,0 +1,410 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamodb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	"github.com/conduitio-labs/conduit-connector-dynamodb/position"
+	"github.com/conduitio/conduit-commons/opencdc"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/google/uuid"
+	"github.com/matryer/is"
+)
+
+var (
+	PartitionKey = "pkey"
+	SortKey      = "skey"
+)
+
+// Records is a slice of opencdc.Record, that can be sorted by the sort key under record.Key["skey"].
+type Records []opencdc.Record
+
+// Implementing the sort interface.
+func (r Records) Len() int {
+	return len(r)
+}
+
+func (r Records) Less(i, j int) bool {
+	// Ensure that both records have a sort key
+	sortKeyI, okI := r[i].Key.(opencdc.StructuredData)[SortKey].(string)
+	sortKeyJ, okJ := r[j].Key.(opencdc.StructuredData)[SortKey].(string)
+
+	// If either key is not present or not a string
+	if !okI || !okJ {
+		return false
+	}
+
+	// Compare the sort key values
+	return sortKeyI < sortKeyJ
+}
+
+func (r Records) Swap(i, j int) {
+	r[i], r[j] = r[j], r[i]
+}
+
+func TestSource_SuccessfulSnapshot(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	client, cfg := prepareIntegrationTest(ctx, t)
+
+	testTable := cfg[SourceConfigTable]
+	source := &Source{}
+	err := source.Configure(context.Background(), cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// insert 5 rows
+	err = insertRecord(ctx, client, testTable, 0, 4)
+	is.NoErr(err)
+
+	err = source.Open(context.Background(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var got Records
+	for {
+		rec, err := source.Read(ctx)
+		if errors.Is(err, sdk.ErrBackoffRetry) {
+			break
+		}
+		got = append(got, rec)
+	}
+	is.True(got != nil)
+	is.Equal(5, got.Len())
+	// sort the records then assert the values.
+	sort.Sort(got)
+	for i, rec := range got {
+		is.Equal(rec.Payload.After, opencdc.StructuredData{PartitionKey: fmt.Sprintf("pkey%d", i), SortKey: fmt.Sprintf("%d", i)})
+	}
+	_ = source.Teardown(ctx)
+}
+
+func TestSource_SnapshotRestart(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	client, cfg := prepareIntegrationTest(ctx, t)
+	testTable := cfg[SourceConfigTable]
+	source := &Source{}
+	err := source.Configure(context.Background(), cfg)
```

Comment:
There's a bunch of `context.Background()` calls in these tests, which can be replaced with `ctx`.